presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/Rectangle.java

```diff
@@ -0,0 +1,147 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.geospatial;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public final class Rectangle
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(Rectangle.class).instanceSize();
+
+    private final double xMin;
+    private final double yMin;
+    private final double xMax;
+    private final double yMax;
+
+    @JsonCreator
+    public Rectangle(
+            @JsonProperty("xmin") double xMin,
+            @JsonProperty("ymin") double yMin,
+            @JsonProperty("xmax") double xMax,
+            @JsonProperty("ymax") double yMax)
+    {
+        checkArgument(xMin <= xMax, "xMin is greater than xMax");
+        checkArgument(yMin <= yMax, "yMin is greater than yMax");
+        this.xMin = xMin;
+        this.yMin = yMin;
+        this.xMax = xMax;
+        this.yMax = yMax;
+    }
+
+    @JsonProperty
+    public double getXMin()
+    {
+        return xMin;
+    }
+
+    @JsonProperty
+    public double getYMin()
+    {
+        return yMin;
+    }
+
+    @JsonProperty
+    public double getXMax()
+    {
+        return xMax;
+    }
+
+    @JsonProperty
+    public double getYMax()
+    {
+        return yMax;
+    }
+
+    public double getWidth()
+    {
+        return xMax - xMin;
+    }
+    public double getHeight()
+    {
+        return yMax - yMin;
+    }
+
+    public boolean disjoint(Rectangle other)
+    {
+        requireNonNull(other, "other is null");
+        if (other.xMax < this.xMin || other.xMin > this.xMax) {
+            return true;
+        }
+
+        return (other.yMax < this.yMin || other.yMin > this.yMax);
+    }
+
+    public Rectangle intersection(Rectangle other)
+    {
+        requireNonNull(other, "other is null");
+        if (disjoint(other)) {
+            return null;
+        }
+
+        return new Rectangle(max(this.xMin, other.xMin), max(this.yMin, other.yMin), min(this.xMax, other.xMax), min(this.yMax, other.yMax));
+    }
+
+    public Rectangle merge(Rectangle other)
+    {
+        requireNonNull(other, "other is null");
```

Comment:
This is not really necessary. It will fail with NPE in the line below if it comes to it. In general, it's more important to do these kinds of checks when holding on to the value, which may cause the code to fail much later in an unrelated place.